### PR TITLE
workflows/tests: limit use of `CI-long-timeout` label

### PR DESCRIPTION
**Pull request by carlocab**
The `CI-long-timeout` label is currently overused: it is almost always
added (immediately) to every PR that takes longer than an hour to run, and we suffer
from long CI queues as a consequence. This defeats the purpose of having
the label in the first place.

Let's try to mitigate this by automatically removing the
`CI-long-timeout` label whenever there are at least two open PRs that
already have the label applied.

I employ a little hack to be able to use our `label-pull-requests`
action, since trying to do this with `actions/github-script` is a little
involved. An alternative to this is to use an external action such as
`andymckay/labeler`. [1] I can also look at adapting our action
properly, but this might take a while since I don't know JavaScript.

[1] https://github.com/andymckay/labeler

### CODE REVIEW

File: .github/workflows/tests.yml
Comment (carlocab) [inline .github/workflows/tests.yml:72]: Ah, whoops. I don't think this token is accessible here. But then this probably needs a token with write access to pull requests.

Comment (carlocab) [reply .github/workflows/tests.yml:72]: One way to get around this is to skip removing the label and just refusing to set a long timeout if there are too many pull requests with the label, but that's bound to make using the label very confusing.

**Review by MikeMcQuaid — DISMISSED** [review]
This makes sense to me. I'd love to get this logic out of the YAML file in the medium-term but this fixes a problem we have today.

Another, perhaps simpler, approach would be to just immediately fail CI if `CI-long-timeout` is specified and there's too many PRs using it. That'd need the maintainer to change labels and rerun the build.

**Comment by carlocab** [comment]
> Another, perhaps simpler, approach would be to just immediately fail CI if `CI-long-timeout` is specified and there's too many PRs using it. That'd need the maintainer to change labels and rerun the build.

Oh, I like this idea. Especially since it gets around the issue of having access to a token with the right permissions.

**Comment by carlocab** [comment]
Ok, this works. Applying the `CI-long-timeout` label to this PR results in the error
```
Error: Too many pull requests (31) with the long-timeout label!
```
https://github.com/Homebrew/homebrew-core/runs/4457463863?check_suite_focus=true#step:2:34

File: .github/workflows/tests.yml
Comment (carlocab) [inline .github/workflows/tests.yml:48]: Except this doesn't do what I thought it would do, which is just run this check only when the PR itself has the `CI-long-timeout` label.

File: .github/workflows/tests.yml
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:75]: Perhaps this could be part of `tap_syntax` as it's already doing label handling?

Comment (carlocab) [reply .github/workflows/tests.yml:75]: Yep, makes sense. I'd been focusing on getting something working first before trying to figure out how to integrate this into the rest of the workflow.

**Comment by carlocab** [comment]
Applying `CI-long-timeout` here produces an error... https://github.com/Homebrew/homebrew-core/runs/4457854232?check_suite_focus=true#step:2:109

File: .github/workflows/tests.yml
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:127]:
```suggestion
              const long_pr_count = ${{ needs.tap_syntax.outputs.long_pr_count }}
              const maximum_long_pr_count = 2
              if (long_pr_count > maximum_long_pr_count) {
                core.setFailed('Too many pull requests (${long_pr_count} > ${maximum_long_pr_count}) with the long-timeout label! Remove the label from this or some of those PRs, once their CI has completed).')
              }
```

File: .github/workflows/tests.yml
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:50]: Wonder if a future extension could be to ensure that we only apply this on PRs where the CI is already running?

**Comment by carlocab** [comment]
> Applying `CI-long-timeout` here produces an error... https://github.com/Homebrew/homebrew-core/runs/4457854232?check_suite_focus=true#step:2:109

And omitting it elides the error. Ok, everything working.

File: .github/workflows/tests.yml
Comment (carlocab) [reply .github/workflows/tests.yml:50]: That would be nice, yes. But I'd rather leave that to someone who understands GitHub's GraphQL API better than I do. (This could very well be future-me.)

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by Bo98** [comment]
> I'd love to get this logic out of the YAML file in the medium-term but this fixes a problem we have today.

The medium term fix will be to roll this into ephemeral scheduling.

**Comment by XuehaiPan** [comment]
Can we automatically add a label `CI-failed-with-timeout` (or something similar) to the pull request when the test-bot failed with running timeout?

The logic may be:

```
on test start:
    if labeled with `CI-failed-with-timeout`
        remove label `CI-failed-with-timeout`

on test failed:
    if failed with `CI workflow run canceled` (need to handle force push, that can also cancel the workflow)
        add label `CI-failed-with-timeout`
```

**Comment by SMillerDev** [comment]
That doesn't really work because it doesn't always time out because it has a lot of things to do. It can also time out because something is stuck. And automatically adding a label won't fix that.

**Comment by MikeMcQuaid** [comment]
I think a label like that would be good and helpful. I wouldn't propose we do any different behaviour based on it, though 👍🏻

**Pull request merged**
